### ISSUE TITLE
Assertion failure in mkdir_p_at_internal()

### ISSUE THREAD

**Issue by pwithnall** [ostreedev/ostree#1215]
We got a failure report from an EOS machine which included the following line:
```
OSTree:ERROR:libglnx/glnx-shutil.c:215:mkdir_p_at_internal: assertion failed: (lastslash != NULL)
```

There’s not any other information from the report to go on, but obviously it’s possible for that assertion to fail. I wonder if `mkdirat()` can return `ENOENT` if part of the `dirfd` path is deleted (after the `dirfd` is opened but before it’s passed to `mkdirat()`)?

**Comment by pwithnall** [comment]
> I wonder if `mkdirat()` can return `ENOENT` if part of the `dirfd` path is deleted (after the `dirfd` is opened but before it’s passed to `mkdirat()`)?

Yup, that causes an assertion failure. Reproducible with the following test case (I can’t easily put this in libglnx because it doesn’t have any test infrastructure and I don’t have time to add any).

```
static void
test_glnx_mkdir_p (void)
{
  int dfd;
  g_autoptr(GError) error = NULL;

  g_assert (glnx_ensure_dir (AT_FDCWD, "test", 0755, NULL));
  g_assert (glnx_opendirat (AT_FDCWD, "test", FALSE, &dfd, NULL));

  rmdir ("test");

  glnx_shutil_mkdir_p_at (dfd, "blah/baz", 0755, NULL, &error);
  g_assert_no_error (error);
}
```

**Comment by pwithnall** [comment]
Opened https://github.com/GNOME/libglnx/pull/87.

**Comment by pwithnall** [comment]
Fixed with commit e30154431d7eea6397e5502b175ba3b50330140f.
